core/src/types/list.rs

```diff
@@ -80,44 +92,77 @@ impl Stream for Lister {
             return Poll::Ready(None);
         }
 
-        if let Some(fut) = self.stating.as_mut() {
-            let (path, rp) = ready!(fut.poll_unpin(cx));
-
-            // Make sure we will not poll this future again.
-            self.stating = None;
-            let metadata = match rp {
-                Ok(rp) => rp.into_metadata(),
-                Err(err) => {
+        if self.listing && self.task_queue.len() < self.task_queue.capacity() {
```

Comment:
We can use `Option<Lister>` to make it more clear that the lister has been consumed.

```rust
if let Some(lister) = self.lister.as_mut() {
   if self.task_queue.len() < self.task_queue.capacity() {
       match self.lister.poll_next(cx) {
           Poll::Pending => {},
           Poll::Ready(Ok(Some(oe))) => {
               // build stat task
               self.task_queue.push_back(task);
           }
           Poll::Ready(Ok(None)) => { self.lister = None; }
           Poll::Ready(Ok(Err(err))) => { .. }
       }     
   }
}

if let Some(task) = self.task_queue.front_mut() {
    return match task {
        StatTask::Handle(handle) => {
            ...
            return Poll::Ready(Some(Ok(Entry::new(path, metadata))))
        }
        StatTask::KnownEntry(entry) => {
            return Poll::Ready(Some(Ok(Entry::new(path, metadata))))
        }
    }
}
```

We don't keep look calling `poll_next`. I think this can make code more easy to read.

Reply:
Adopting `Option<Lister>` is a very wise solution. I learned a lot. Thank you! 👍